src/nxt_h1proto.c

```diff
@@ -872,52 +873,39 @@ nxt_h1p_request_body_read(nxt_task_t *task, nxt_http_request_t *r)
     switch (h1p->transfer_encoding) {
 
     case NXT_HTTP_TE_CHUNKED:
-        status = NXT_HTTP_LENGTH_REQUIRED;
-        goto error;
+        r->chunked = 1;
+        h1p->chunked_parse.mem_pool = r->mem_pool;
+        break;
 
     case NXT_HTTP_TE_UNSUPPORTED:
-        status = NXT_HTTP_NOT_IMPLEMENTED;
-        goto error;
+        h1p->keepalive = 0;
+        nxt_http_request_error(task, r, NXT_HTTP_NOT_IMPLEMENTED);
+        return;
 
     default:
     case NXT_HTTP_TE_NONE:
         break;
     }
 
-    if (r->content_length_n == -1 || r->content_length_n == 0) {
+    if ((r->content_length_n <= 0) && !r->chunked) {
         goto ready;
     }
 
-    body_length = (size_t) r->content_length_n;
+    body_length = (ssize_t) r->content_length_n;
```

Comment:
This cast shouldn't be needed.

body_length is a `ssize_t` and  r->content_length_n is a `nxt_off_t` aka `off_t` which is defined as signed by [POSIX](https://pubs.opengroup.org/onlinepubs/009696899/basedefs/sys/types.h.html)

So both types are `signed`, the only difference being on 32bit, ssize_t is 4 bytes whereas off_t is still 8 bytes due to _FILE_OFFSET_BITS=64